perses/dispersed/utils.py

```diff
@@ -0,0 +1,705 @@
+    sampler_state : openmmtools.states.SamplerState
+        The starting state at which to minimize the system.
+    max_iterations : int, optional, default 20
+        The maximum number of minimization steps. Default is 100.
+
+    Returns
+    -------
+    sampler_state : openmmtools.states.SamplerState
+        The posititions and accompanying state following minimization
+    """
+    if type(cache.global_context_cache) == cache.DummyContextCache:
+        integrator = openmm.VerletIntegrator(1.0) #we won't take any steps, so use a simple integrator
+        context, integrator = cache.global_context_cache.get_context(thermodynamic_state, integrator)
+        _logger.debug(f"using dummy context cache")
+    else:
+        _logger.debug(f"using global context cache")
+        context, integrator = cache.global_context_cache.get_context(thermodynamic_state)
+    sampler_state.apply_to_context(context, ignore_velocities = True)
+    openmm.LocalEnergyMinimizer.minimize(context, maxIterations = max_iterations)
+    sampler_state.update_from_context(context)
+
+def multinomial_resample(total_works, num_resamples):
+    """
+    from a numpy array of total works and particle_labels, resample the particle indices N times with replacement
+    from a multinomial distribution conditioned on the weights w_i \propto e^{-cumulative_works_i}
+    Parameters
+    ----------
+    total_works : np.array of floats
+        generalized accumulated works at time t for all particles
+    num_resamples : int, default len(sampler_states)
+        number of resamples to conduct; default doesn't change the number of particles
+
+    Returns
+    -------
+    resampled_works : np.array([1.0/num_resamples]*num_resamples)
+        resampled works (uniform)
+    resampled_indices : np.array of ints
+        resampled indices
+    """
+    normalized_weights = np.exp(-total_works - logsumexp(-total_works))
+    resampled_indices = np.random.choice(len(normalized_weights), num_resamples, p=normalized_weights, replace = True)
+    resampled_works = np.array([np.average(total_works)] * num_resamples)
+
+    return resampled_works, resampled_indices
+
+def ESS(works_prev, works_incremental):
+    """
+    compute the effective sample size (ESS) as given in Eq 3.15 in https://arxiv.org/abs/1303.3123.
+    Parameters
+    ----------
+    works_prev: np.array
+        np.array of floats representing the accumulated works at t-1 (unnormalized)
+    works_incremental: np.array
+        np.array of floats representing the incremental works at t (unnormalized)
+
+    Returns
+    -------
+    ESS: float
+        effective sample size
+    """
+    prev_weights_normalized = np.exp(-works_prev - logsumexp(-works_prev))
+    incremental_weights_unnormalized = np.exp(-works_incremental)
+    ESS = np.dot(prev_weights_normalized, incremental_weights_unnormalized)**2 / np.dot(np.power(prev_weights_normalized, 2), np.power(incremental_weights_unnormalized, 2))
```

Comment:
slightly hard to read, could this be split into two or three lines?

Reply:
Is there a reason that `**2` is used on the numerator and `np.power(x,2)` used in the denomentor?